##QTM350 HW2 Sizhe Ren & Qiao Lin
#Introducing some fundamental methods from classic SKlearn package
What is SKlearn? A machine learning package based on  NumPy, SciPy, and matplotlib. Now we are going to introduce this package to you by showing you a  Decision Tree based classification algorithm provide by Sklearn package.



Installing the package

In [0]:
pip install -U scikit-learn

We are using the default dataset iris imbedded in this package. The dataset itself is fairly simple: each tuple includes four numerical values. Our job is to classify different kinds of iris flower, Iris Setosa, Iris Versicolour and Iris Virginica by these four attributes. Each attribute corresponds to sepal length, sepal width, petal length and petal width. 

In [0]:
from  sklearn import datasets
iris = datasets.load_iris()
x = iris.data
y = iris.target

Splitting the dataset:

The iris dataset is provided by sklearn so we can download it directly from the package. Then we use the function train_test_split provided by model_selection package to separate our data into training data and test data. This function acceipt a parameter called test_size which specify the size of testset. Here we set it to 0.5 so that train set and testset are equally separated.

In [0]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=.5)

Build the model:

To use the model classifier provided by Sklearn, we import the tree classifier. This is a Decision tree based classifier that should do a good work classifying different iris species based on features. We also tried another classifier based on KNN which should also do the work. We want to compare the result later.

In [0]:
from sklearn import tree
classifier1=tree.DecisionTreeClassifier()

In [0]:
from sklearn import neighbors
classifier2=neighbors.KNeighborsClassifier()

Train the Model.

The fit function here is provided by Sklearn, it trains the training data according to the classifiers. For example, in the code below the training data is first trained with Decision Tree based model and then trained with KNN based model. Now we have completed the training process using Sklearn provided packages.

In [0]:
classifier1.fit(x_train,y_train)

In [0]:
classifier2.fit(x_train,y_train)

Make predictions:

The predict function is also provided by Sklearn. It makes prediction based on data in x_test. Here, for each data in testset, this function make a prediction of its iris species based on features and model and save the predictions into a list called predictions.

Then we use another package provided by sklearn.metrics called accuracy_score. This function compares each guessing resulting saved in predictions with the true value in y_test and see how many of them are predicted correctly. The output is the percentage of correctly matching data.

In [0]:
predictions=classifier1.predict(x_test)
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,predictions))

0.9733333333333334


In [0]:
predictions=classifier2.predict(x_test)
from sklearn.metrics import accuracy_score
print(accuracy_score(y_test,predictions))

Output:

The tree based classifier reach an accuracy of 97.33% while the KNN based classifier is 96%. So here the tree based model did a slightly better job on predicting iris species type.

All codes and this file can be found on Github:
https://github.com/QiaoLin22/QTM350